torch.flip via TI

lezcano opened the pull request:
Implements an idea by @ngimel to improve the performance of `torch.flip` via a clever hack into TI to bypass the fact that TI is not designed to work with negative indices.

Something that might be added is vectorisation support on CPU, given how simple the implementation is now.

Some low-hanging fruits that I did not implement:
- Write it as a structured kernel
- Migrate the tests to opinfos
- Have a look at `cumsum_backward` and `cumprod_backward`,  as I think that they could be implemented faster with `flip`, now that `flip` is fast.

**Edit**
This operation already has OpInfos and it cannot be migrated to a structured kernel because it implements quantisation

Summary of the PR:
- x1.5-3 performance boost on CPU
- x1.5-2 performance boost on CUDA
- Comparable performance across dimensions, regardless of the strides (thanks TI)
- Simpler code


<details>
<summary>
Test Script
</summary>

```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, dims, num_threads, device):
    x = torch.rand(*size, device=device)

    timer = Timer(
        "torch.flip(x, dims=dims)",
        globals={"x": x, "dims": dims},
        label=f"Flip {device}",
        description=f"dims: {dims}",
        sub_label=f"size: {size}",
        num_threads=num_threads,
    )

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    sizes = ((1000,)*2, (1000,)*3, (10000,)*2)
    for size, device in product(sizes, ("cpu", "cuda")):
        threads = (1, 2, 4) if device == "cpu" else (1,)
        list_dims = [(0,), (1,), (0, 1)]
        if len(size) == 3:
            list_dims.append((0, 2))
        for num_threads, dims in product(threads, list_dims):
            yield size, dims, num_threads, device


def compare():
    compare = Compare([get_timer(*params) for params in get_params()])
    compare.trim_significant_figures()
    compare.colorize()
    compare.print()


compare()
```
</details>

<details>
<summary>
Benchmark PR
</summary>

![image](https://user-images.githubusercontent.com/3291265/119139954-81e46d80-ba3b-11eb-9aad-e825e515d41b.png)

</details>

<details>
<summary>
Benchmark master
</summary>

![image](https://user-images.githubusercontent.com/3291265/119139915-76914200-ba3b-11eb-9aa8-84b3ca220c93.png)

</details>

Fixes #16424

lezcano committed:
Initial commit

lezcano committed:
Working CPU/CUDA implementation

lezcano committed:
Make opaque

lezcano committed:
Remove flip_strides

lezcano committed:
Remove quantized for now

ngimel commented:
ROCm error is real 
```
Could not run 'aten::flip' with arguments from the 'QuantizedCPU' backend.
```

ngimel commented:
@zasdfgbnm fyi - this PR adds signed strides template type to OffsetCalculator

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -1,159 +1,71 @@
+  auto flip_dims_b = at::dim_list_to_bitset(dims, total_dims);
+  Tensor out_tensor = at::empty_like(in_tensor, MemoryFormat::Contiguous);
+  int n = 0;
+  auto strides = out_tensor.strides().vec();
```
Using DimVector for strides is slightly faster than std::vector

Lezcano replied:
Done. The question is then, why does this method return a `std::vector` by default rather than a `DimVector`?

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -1,159 +1,71 @@
+
+  auto* data = reinterpret_cast<char*>(iter.data_ptr(0));
+  const auto sizes = iter.shape();
+  auto strides_bytes = iter.strides(0).vec();
```
similarly here, you can avoid std::vector

ngimel commented on aten/src/ATen/TensorIterator.h:
```diff
@@ -313,6 +313,10 @@ struct TORCH_API TensorIteratorBase : public impl::MetaBase {
   PtrVector get_base_ptrs() const;
 
+  // Helper functions for advanced stride manipulations (e.g. torch.flip)
+  void set_arg_strides(const int arg, IntArrayRef strides) { operands_[arg].stride_bytes = std::move(strides); }
```
Probably makes sense to call these functions `_unsafe_set_arg_strides`, and `_unsafe_set_arg_data`, because no checks are made to validate the data

ngimel reviewed (commented):
This looks good, merge conflict needs to be solved, and quantization dispatch needs to be added back. Quantization can be handled in the cpu stub, the way it was before, and looks like we actually have tests that would trigger that :-) You could also add directed quantization test if you want.

ngimel commented on aten/src/ATen/cuda/detail/OffsetCalculator.cuh:
```diff
@@ -2,6 +2,7 @@
 
```
this is a very surgical change that allows signed strides, slick approach!

ngimel commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -458,11 +459,46 @@ Tensor & masked_scatter__cuda(Tensor& self, const Tensor& mask, const Tensor& so
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16,
+                                         iter.dtype(), "flip_cuda",
+  [&] {
+    using dtype = OpaqueType<sizeof(scalar_t)>;
```
Thanks for using OpaqueType here :-)

ngimel commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -458,11 +459,46 @@ Tensor & masked_scatter__cuda(Tensor& self, const Tensor& mask, const Tensor& so
+template <typename scalar_t>
+void flip_kernel_impl(TensorIterator& iter) {
+  if (!iter.can_use_32bit_indexing()) {
+    for (auto& sub_iter : iter.with_32bit_indexing()) {
```
`with_32_bit_indexing()` is using `get_dim_to_split` function https://github.com/pytorch/pytorch/blob/9fc0c5a54a17689dd92029347473df69ece1f409/aten/src/ATen/TensorIterator.cpp#L1087 that, as you can see, won't work correctly with negative strides. 
So there are 2 options here: think about how it can be modified, or instantiate 64-bit kernel and offset calculator for large tensors. Unfortunately, we don't routinely run tests for large tensors, so this was not flagged.

lezcano committed:
Addresss Natalia's comments

Lezcano commented:
@ngimel I addressed the comments. Let's see what the CI has to say this time. If it does not complain, the code should be ready for another review.

A number of things:
- I added the `QuantizedCPU` dispatch, but I have not been able to test the implementation properly. It seems like tests and support of quantised methods is lacking, so I have not been able to find examples on how to create quantised tensors. I'd appreciate pointers in this direction.
- I've seen that there's a `QuantizedCUDA` dispatch, but I do not know what's the deal with that one. 
- I patched CUDA for large tensors, but they are not tested yet. Should I add some tests comparing torch cuda vs numpy (cpu)? Can numpy handle those kind of tensors? Otherwise I could add the poorman's version of this test, doing `a[i] == a.flip(0)[-i-1]` for a number of randomly sampled `i`'s

lezcano committed:
Add vectorized implementation

Lezcano commented:
I have also just added the vectorised version. This further improves the speed of the operation in some cases (when `dim=0`). I do not know why does it improve the performance in some cases. I cannot explain why we see this performance boost in just some cases. In particular, it improves much more the performance for the 3D tensor than for the 2D tensor. Do you know what could be causing this @ngimel ?

<details>
<summary>
Vectorization
</summary>

```
[----------------------------------------- Flip cpu -----------------------------------------]
                                |  dims: (0,)  |  dims: (1,)  |  dims: (0, 1)  |  dims: (0, 2)
1 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |        94    |       359    |        346     |
      size: (1000, 1000, 1000)  |    950000    |    960000    |     959000     |    1300000
      size: (10000, 10000)      |    105000    |    126600    |     129000     |
2 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |        48    |       185    |        179     |
      size: (1000, 1000, 1000)  |    685500    |    690000    |     690000     |     832000
      size: (10000, 10000)      |     74100    |     84300    |      85200     |
4 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |        30    |        97    |         94     |
      size: (1000, 1000, 1000)  |    480000    |    480000    |     480000     |     530000
      size: (10000, 10000)      |     51900    |     55200    |      55700     |

Times are in microseconds (us)
```

</details>

<details>
<summary>
No Vectorization
</summary>

```
[----------------------------------------- Flip cpu -----------------------------------------]
                                |  dims: (0,)  |  dims: (1,)  |  dims: (0, 1)  |  dims: (0, 2)
1 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |       354    |       352    |        345     |
      size: (1000, 1000, 1000)  |   1170000    |   1260000    |    1260000     |    1260000
      size: (10000, 10000)      |    127000    |    127000    |     129000     |
2 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |       181    |       181    |        178     |
      size: (1000, 1000, 1000)  |    810000    |    852000    |     851000     |     821000
      size: (10000, 10000)      |     87400    |     84600    |      85800     |
4 threads: -----------------------------------------------------------------------------------
      size: (1000, 1000)        |        96    |        96    |         94     |
      size: (1000, 1000, 1000)  |    537000    |    556000    |     555000     |     523000
      size: (10000, 10000)      |     58600    |     55500    |      56180     |

Times are in microseconds (us).
```

</details>

nb. the timings differ from those in the OP because I run the tests in a different box

ngimel commented:
To prepare outputs for quantized tests, you can use an approach similar to one of the tests in test_torch
https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_torch.py#L2731-L2734
Numpy is able to handle large tensors, so you can compare with numpy (there's a `compare_with_numpy` utility function for that). However, you'd need at least 2 GB input tensor and 2 GB output tensor, so you'd need to guard a test with some @largeTensorTest decorator.

ngimel commented:
FPE errors in the tests are real.

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -1,159 +1,72 @@
+  auto in_tensor = self;
+  const int64_t total_dims = in_tensor.dim();
+  auto flip_dims_b = at::dim_list_to_bitset(dims, total_dims);
+  Tensor out_tensor = at::empty_like(in_tensor, MemoryFormat::Contiguous);
```
please don't hardcode output layout here, generally operations try to propagate the permutation of the input tensor.

Lezcano replied:
Oh, what's the reason for this? I thought that contiguous was always better.

Lezcano replied:
Makes sense, thanks!

lezcano committed:
Fix bug when the tensor has zero elements
Write OpInfo as generator
MemoryFormat::Preserve
More const

lezcano committed:
More thorough testing
Added quantized testing
Added large tensor testing

Lezcano commented:
@ngimel Fixed the bug and implemented the tests. I could not run the largeTensor tests locally, so let's see what the CI has to say.

lezcano committed:
Flake

Lezcano commented:
`test_flip_numpy_cpu_bool` error seems real. I'll look into it tomorrow.

lezcano committed:
More thorough flip testing
compare_with_numpy now works with bfloat16

lezcano committed:
Include more numpy tests
Deactivate bool
Simpler shape generator

Lezcano commented:
@ngimel I added more thorough testing, but asan still complains on `flip_cpu_bool` with the following error:
<details>
<summary>
Scary Error
</summary>

```
Jun 02 11:00:52   test_flip_numpy_cpu_bool (__main__.TestShapeOpsCPU) ... /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:115:17: runtime error: load of value 190, which is not a valid value for type 'bool'
Jun 02 11:00:53     #0 0x7faa2b1abd47 in std::_Tuple_impl<0ul, bool, bool>::_Tuple_impl<bool&, bool&, void>(bool&, bool&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xdc12d47)
Jun 02 11:00:53     #1 0x7faa35af6c58 in void at::native::(anonymous namespace)::basic_loop<at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}>(char* restrict*, long const*, long, long, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}&&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x1855dc58)
Jun 02 11:00:53     #2 0x7faa35af62df in void at::native::(anonymous namespace)::vectorized_loop<at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}>(char**, long, long, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}&&, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}&&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x1855d2df)
Jun 02 11:00:53     #3 0x7faa35af57c3 in void c10::function_ref<void (char**, long const*, long, long)>::callback_fn<auto at::TensorIteratorBase::loop_2d_from_1d<void at::native::(anonymous namespace)::cpu_kernel_vec<true, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}>(at::TensorIteratorBase&, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}&&, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}&&)::{lambda(char**, long const*, long)#1}>(void at::native::(anonymous namespace)::cpu_kernel_vec<true, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}>(at::TensorIteratorBase&, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(bool, bool)#1}&&, at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const::{lambda()#20}::operator()() const::{lambda(at::vec::(anonymous namespace)::Vectorized<bool>, at::vec::(anonymous namespace))#1}&&)::{lambda(char**, long const*, long)#1} const&)::{lambda(char**, long const*, long, long)#1}>(long, char**, long const*, long, long) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x1855c7c3)
Jun 02 11:00:53     #4 0x7faa2a4e1077 in at::internal::serial_for_each(c10::ArrayRef<long>, c10::ArrayRef<long>, char**, unsigned long, c10::function_ref<void (char**, long const*, long, long)>, at::Range) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xcf48077)
Jun 02 11:00:53     #5 0x7faa2a4ca36a in at::TensorIteratorBase::serial_for_each(c10::function_ref<void (char**, long const*, long, long)>, at::Range) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xcf3136a)
Jun 02 11:00:53     #6 0x7faa2a4c9693 in at::TensorIteratorBase::for_each(c10::function_ref<void (char**, long const*, long, long)>, long) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xcf30693)
Jun 02 11:00:53     #7 0x7faa35ade611 in at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool)::$_10::operator()() const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x18545611)
Jun 02 11:00:53     #8 0x7faa35adaaa0 in at::native::(anonymous namespace)::flip_kernel(at::TensorIterator&, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x18541aa0)
Jun 02 11:00:53     #9 0x7faa2af30345 in at::native::flip(at::Tensor const&, c10::ArrayRef<long>) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xd997345)
Jun 02 11:00:53     #10 0x7faa2c66fd84 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::ArrayRef<long>), &at::(anonymous namespace)::(anonymous namespace)::wrapper__flip>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::ArrayRef<long> > >, at::Tensor (at::Tensor const&, c10::ArrayRef<long>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xf0d6d84)
Jun 02 11:00:53     #11 0x7faa2c35e7a4 in at::Tensor c10::Dispatcher::redispatch<at::Tensor, at::Tensor const&, c10::ArrayRef<long> >(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::ArrayRef<long>)> const&, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xedc57a4)
Jun 02 11:00:53     #12 0x7faa2c1e2951 in at::redispatch::flip(c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xec49951)
Jun 02 11:00:53     #13 0x7faa3153ac9b in torch::autograd::VariableType::(anonymous namespace)::flip(c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x13fa1c9b)
Jun 02 11:00:53     #14 0x7faa3153a520 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>), &torch::autograd::VariableType::(anonymous namespace)::flip>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long> > >, at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<long>) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x13fa1520)
Jun 02 11:00:53     #15 0x7faa2d1db49b in at::Tensor::flip(c10::ArrayRef<long>) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xfc4249b)
Jun 02 11:00:53     #16 0x7faa4f62fa23 in torch::autograd::THPVariable_flip(_object*, _object*, _object*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x383ba23)
Jun 02 11:00:53     #17 0x558240889ed3 in _PyCFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/methodobject.c:231
Jun 02 11:00:53     #18 0x55824088a2ee in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2313
Jun 02 11:00:53     #19 0x55824097d1f1 in partial_call /tmp/build/80754af9/python_1614113050744/work/Modules/_functoolsmodule.c:183
Jun 02 11:00:53     #20 0x55824088a0ba in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2331
Jun 02 11:00:53     #21 0x558240911c4d in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4875
Jun 02 11:00:53     #22 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #23 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #24 0x55824090bc60 in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4992
Jun 02 11:00:53     #25 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #26 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #27 0x55824090c778 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #28 0x55824090c778 in PyEval_EvalCodeEx /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4187
Jun 02 11:00:53     #29 0x55824090d593 in function_call /tmp/build/80754af9/python_1614113050744/work/Objects/funcobject.c:604
Jun 02 11:00:53     #30 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #31 0x558240935951 in do_call_core /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5120
Jun 02 11:00:53     #32 0x558240935951 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3404
Jun 02 11:00:53     #33 0x55824090b0cd in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #34 0x55824090bc60 in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4992
Jun 02 11:00:53     #35 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #36 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #37 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #38 0x55824090bc60 in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4992
Jun 02 11:00:53     #39 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #40 0x55824093508a in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3351
Jun 02 11:00:53     #41 0x55824090af65 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #42 0x55824090bc60 in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4992
Jun 02 11:00:53     #43 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #44 0x55824093508a in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3351
Jun 02 11:00:53     #45 0x55824090af65 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #46 0x55824090c357 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #47 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #48 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #49 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #50 0x558240935951 in do_call_core /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5120
Jun 02 11:00:53     #51 0x558240935951 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3404
Jun 02 11:00:53     #52 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #53 0x55824090c13d in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #54 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #55 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #56 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #57 0x5582408e2970 in slot_tp_call /tmp/build/80754af9/python_1614113050744/work/Objects/typeobject.c:6207
Jun 02 11:00:53     #58 0x55824088a0ba in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2331
Jun 02 11:00:53     #59 0x558240911c4d in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4875
Jun 02 11:00:53     #60 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #61 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #62 0x55824090c357 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #63 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #64 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #65 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #66 0x558240935951 in do_call_core /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5120
Jun 02 11:00:53     #67 0x558240935951 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3404
Jun 02 11:00:53     #68 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #69 0x55824090c13d in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #70 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #71 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #72 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #73 0x5582408e2970 in slot_tp_call /tmp/build/80754af9/python_1614113050744/work/Objects/typeobject.c:6207
Jun 02 11:00:53     #74 0x55824088a0ba in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2331
Jun 02 11:00:53     #75 0x558240911c4d in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4875
Jun 02 11:00:53     #76 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #77 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53  
   #78 0x55824090c357 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #79 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #80 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #81 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #82 0x558240935951 in do_call_core /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5120
Jun 02 11:00:53     #83 0x558240935951 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3404
Jun 02 11:00:53     #84 0x55824090adc3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #85 0x55824090c13d in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #86 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #87 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #88 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #89 0x5582408e2970 in slot_tp_call /tmp/build/80754af9/python_1614113050744/work/Objects/typeobject.c:6207
Jun 02 11:00:53     #90 0x55824088a0ba in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2331
Jun 02 11:00:53     #91 0x558240911c4d in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4875
Jun 02 11:00:53     #92 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #93 0x55824090ba2a in _PyFunction_FastCall /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4933
Jun 02 11:00:53     #94 0x55824090ba2a in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4968
Jun 02 11:00:53     #95 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #96 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #97 0x55824090ba2a in _PyFunction_FastCall /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4933
Jun 02 11:00:53     #98 0x55824090ba2a in fast_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4968
Jun 02 11:00:53     #99 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #100 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #101 0x55824090b270 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #102 0x55824090c357 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:5084
Jun 02 11:00:53     #103 0x55824088a29e in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2310
Jun 02 11:00:53     #104 0x55824088eda2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2373
Jun 02 11:00:53     #105 0x558240889cdd in PyObject_Call /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2261
Jun 02 11:00:53     #106 0x5582408e1fea in slot_tp_init /tmp/build/80754af9/python_1614113050744/work/Objects/typeobject.c:6420
Jun 02 11:00:53     #107 0x558240911e36 in type_call /tmp/build/80754af9/python_1614113050744/work/Objects/typeobject.c:915
Jun 02 11:00:53     #108 0x55824088a0ba in _PyObject_FastCallDict /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2331
Jun 02 11:00:53     #109 0x55824090bf49 in _PyObject_FastCallKeywords /tmp/build/80754af9/python_1614113050744/work/Objects/abstract.c:2496
Jun 02 11:00:53     #110 0x558240911c4d in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4875
Jun 02 11:00:53     #111 0x55824093508a in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3351
Jun 02 11:00:53     #112 0x5582408b7b9a in _PyFunction_FastCall /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4933
Jun 02 11:00:53     #113 0x558240911bd4 in call_function /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4872
Jun 02 11:00:53     #114 0x5582409342c9 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:3335
Jun 02 11:00:53     #115 0x55824090c778 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4166
Jun 02 11:00:53     #116 0x55824090c778 in PyEval_EvalCodeEx /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:4187
Jun 02 11:00:53     #117 0x55824090d51b in PyEval_EvalCode /tmp/build/80754af9/python_1614113050744/work/Python/ceval.c:731
Jun 02 11:00:53     #118 0x55824098d5e3 in run_mod /tmp/build/80754af9/python_1614113050744/work/Python/pythonrun.c:1025
Jun 02 11:00:53     #119 0x55824098d9e0 in PyRun_FileExFlags /tmp/build/80754af9/python_1614113050744/work/Python/pythonrun.c:978
Jun 02 11:00:53     #120 0x55824098dbe2 in PyRun_SimpleFileExFlags /tmp/build/80754af9/python_1614113050744/work/Python/pythonrun.c:419
Jun 02 11:00:53     #121 0x5582409916d2 in run_file /tmp/build/80754af9/python_1614113050744/work/Modules/main.c:340
Jun 02 11:00:53     #122 0x5582409916d2 in Py_Main /tmp/build/80754af9/python_1614113050744/work/Modules/main.c:811
Jun 02 11:00:53     #123 0x55824085bb0d in main /tmp/build/80754af9/python_1614113050744/work/Programs/python.c:69
Jun 02 11:00:53     #124 0x7faa5835883f in __libc_start_main /build/glibc-S7Ft5T/glibc-2.23/csu/../csu/libc-start.c:291
Jun 02 11:00:53     #125 0x55824093ad6f in _start /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:103
Jun 02 11:00:53 
Jun 02 11:00:53 SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:115:17 in 
```

</details>

I have been twiddling around with the CI container, but I have not been able to find the cause of it. I have deactivated the bool test in that function for now. Do you know what could be causing it? If not, should I just create a new issue?

lezcano committed:
Quantized CUDA

Lezcano commented:
I also added support and tests for `QuantizedCUDA`, as it was enough to add it to the dispatcher

lezcano committed:
remove old comment

lezcano committed:
Fix one test

zasdfgbnm commented on test/test_shape_ops.py:
```diff
@@ -338,86 +338,137 @@ def test_clamp_raises_arg_errors(self, device):
+    @onlyCUDA  # CPU is too slow
+    @largeTensorTest('30GB')
+    @dtypes(torch.int64, torch.double, torch.cdouble)
+    def test_flip_large_tensor(self, device, dtype):
```
Is 30GB GPU memory enough?
sizeof(cdouble) * 2**32 = 64GiB.
I suggest testing only on `torch.uint8`, and initialize the tensor with 
```python
torch.empty(2**32 + 1, dtype=dtype).random_()
```
instead of just empty tensor.

ngimel replied:
Or, if you want to test `torch.double` you can use fewer elements. Splitting happens based on the extent in bytes, not in elements, you can see that even now for any unary op - any input more than 2GB will launch 2 kernels.

zasdfgbnm commented:
Do you have GPUs with large enough memory to run the large tensor test? If not, please ping me when you think this is ready. I want to validate the large tensor test on our 40GB A100 or 80GB A100 to see if it works.

lezcano committed:
Large test edited following Gao

lezcano committed:
Random init in large test

lezcano committed:
back to 2**32, just in case

zasdfgbnm commented on test/test_shape_ops.py:
```diff
@@ -338,86 +338,136 @@ def test_clamp_raises_arg_errors(self, device):
+    @onlyCUDA  # CPU is too slow
+    @largeTensorTest('17GB') # 4 tensors of 4GB (in, out) x (torch, numpy) + 1GB
+    def test_flip_large_tensor(self, device):
+        t_in = torch.empty(2**32 + 1, dtype=toch.uint8).random_()
```
```suggestion
        t_in = torch.empty(2**32 + 1, dtype=torch.uint8).random_()
```

Lezcano replied:
This is embarrassing... (fixed)

ngimel commented:
You could try handling bool inputs with `cpu_kernel` instead of `cpu_kernel_vec`, maybe it'll fix the error?

ngimel commented:
> I have also just added the vectorised version. This further improves the speed of the operation in some cases (when `dim=0`). I do not know why does it improve the performance in some cases. I cannot explain why we see this performance boost in just some cases. In particular, it improves much more the performance for the 3D tensor than for the 2D tensor. Do you know what could be causing this @ngimel ?

When you are flipping the last dimension, you cannot vectorize it, and use basic loop instead (yay, contiguous tests in `isContiguous` are actually correctly handling negative strides and coming up with the right answer!), and your results reflect that.

lezcano committed:
Update test/test_shape_ops.py

Co-authored-by: Gao, Xiang <qasdfgtyuiop@gmail.com>

lezcano committed:
Try to correct for UB san

Lezcano commented:
@ngimel still the same problem:
https://app.circleci.com/pipelines/github/pytorch/pytorch/330466/workflows/6364587a-e49d-4562-9d96-c5e0b95bb191/jobs/13873957

Interestingly enough, the test seems to pass for bfloat16

Lezcano commented:
I've track down the error until these lines:
https://github.com/pytorch/pytorch/blob/c8290955908e83b3dae929f1d29030a3acf1ac99/aten/src/ATen/native/cpu/Loops.h#L46-L53

This function creates a tuple of arguments that it's then forwarded to an `std::apply`-like method. When reinterpreting pointers there are thousands of possible roots for UB, but I do not see how this is one of them.

According to the error a value of "190" is being passed to the tuple constructor, but that means that that value is being returned by dereferencing a pointer to bool. I don't understand what's going on really. I'll try making that cast into a `reinterpret_cast` to see if that makes the sanitiser happy, but I doubt it...

ngimel commented:
Ok, it stems from https://github.com/pytorch/pytorch/pull/58747/files#diff-54b494a4dd0af2160d716378bd5a40e1e4a98c94414901d85adbb6a9ae6dbed2R553-R554, where TI is loading value from the second "input", but second "input" is really output and its memory is uninitialized, so anything can be there. It's no big deal because it's always valid memory, and the value itself is never used, but it throws off asan (because unused value is invalid). 
One way to maybe solve this would be to use input as the second dummy input?

lezcano committed:
Address UB

lezcano committed:
Rollback to dummy_out. Initialise to zero for bools

zasdfgbnm reviewed (commented):
`test_flip_large_tensor_cuda` is passing now

lezcano committed:
Better testing

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -67,7 +69,7 @@ Tensor flip(const Tensor& self, IntArrayRef dims) {
   //   - We iterate in the opposite direction (invert the strides)
   for (int i=0; i<iter.ndim(); i++){
-    if (strides_dummy[i] == 0) {
+    if (strides_dummy[i] == 0 && strides_self[i] != 0) {
```
That's a great idea, can you leave a comment about that?

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -1,159 +1,85 @@
-      : num_indexers(num_indexers),
-        indexers(indexers),
-        indexer_strides(indexer_strides) {}
+  // We initialize it to zeros when it's bool to silence asan, who thinks that we are
```
this comment is stale

lezcano committed:
Fix comments

ngimel commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -1,159 +1,87 @@
-        .add_input(input);
-  for (auto& index : indices) {
-    config.add_input(index);
+  // Nothing to do, we return fast (accounts for cases like size = (3, 0), dim = {0})
+  // This needs to happen after dim_list_to_bitset, which is where the checks are made
+  if (self.numel() <= 1) {
+    return self.clone(MemoryFormat::Preserve);
   }
```
```suggestion

```
